### DISCUSSION TITLE
Server reboots & k8s updates cause corruption of volumes

### DISCUSSION THREAD

**Discussion by Torkolis** [longhorn/longhorn#4164] [Q&A]
Hi does anyone have experience how to handle server reboots and k8s updates in context with stateful apps using longhorn volumes?
I have a single node rke2 cluster with longhorn where I have a simple postgres statefulSet which uses a longhorn volume to store it's database files. I have observed it a little while and seen that whenever I reboot the server or do rke2 update (where rke2-server.service restart is needed, which actually restarts all kubernetes components), the logs of the postgres pod always say that the database was not shut down properly and recovery mode was executed after the kubernetes is up and running again. To mee it seems that the volume actually gets detached from the pod and from the postgres process before it is properly shut down and the pod removed. What could be done about it?

**Comment by derekbit** [comment]
IIUC, although the longhorn volume and the PostgreSQL service are shut down correctly, the PostgreSQL still complains the service is not shut down properly after reboot?

> volume actually gets detached from the pod and from the postgres process before it is properly shut down and the pod removed

Shouldn't the postgres process be stopped before detaching the volume?

**Reply by Torkolis** [reply]
> IIUC, although the longhorn volume and the PostgreSQL service are shut down correctly, the PostgreSQL still complains the service is not shut down properly after reboot?

yes.
> Shouldn't the postgres process be stopped before detaching the volume?

yes that is the Idea, but I am not sure how to make it work, because it is longhorn that does it, but it seems that longhorn does not actually wait for the postgres process to stop and the pod to be removed.

**Reply by derekbit** [reply]
This question is more related to how Kubernetes tears down pods and volumes and how they are ordered.

**Reply by Torkolis** [reply]
I still have this issue, how could I make sure, that when I do a server reboot, the pod and volume do not get separated until the process is properly stopped? Where should I ask about this?

**Reply by derekbit** [reply]
Abnorml or normal reboot?

**Reply by Torkolis** [reply]
just normal reboot, like `sudo reboot now`

**Reply by derekbit** [reply]
Do you drain the node first?

**Reply by Torkolis** [reply]
no, I have a single node cluster, whould draining it work?

**Reply by derekbit** [reply]
Can you scale down the workload to 0 or tear down workload before reboot?

**Reply by Torkolis** [reply]
yes scaling to 0 works. But it is rather a workaround than a solution.

**Reply by derekbit** [reply]
In my opinion, scaling down the non-stateless workload for single node cluster is a correct way.
WDYT? @shuo-wu @c3y1huang @innobead

**Reply by innobead** [reply]
YES, especially since this is a single node, so there is no way to ensure the data intact from other replicas. This is not a workaround actually but an operation practice.

**Reply by Torkolis** [reply]
other replicas? I have a problem with stateful set, it is a 1 to 1 relatonship between pod and volume.
